File: Data Structures & Algorithms/balanced-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # res=0
        # self.node=True
        # def height(curr):
        #     if curr==None:
        #         return 0
        #     left=height(curr.left)
        #     right=height(curr.right)
        #     if abs(left-right)>1:
        #          self.node=False
        #     return 1+max(left,right)
        # height(root)
        # return self.node
        def dfs(root):
            if root==None:
                return [True,0]
            left=dfs(root.left)
            right=dfs(root.right)
            balanced=(left[0] and right[0] and
            abs(left[1]-right[1])<=1)
            height=1+max(left[1],right[1])
            return [balanced,height]
        return dfs(root)[0]

       

File: Data Structures & Algorithms/depth-of-binary-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
    //     queue<TreeNode*> q;
    //     if (root!=nullptr){
    //         q.push(root);
    //     }
    //     int level=0;
    //     while(!q.empty()){
    //         int size=q.size();
    //         for(int i=0;i<size;i++){
    //               TreeNode* node=q.front();
    //             q.pop();
    //             if(node->left!=nullptr){
    //                 q.push(node->left);
    //             }
    //             if(node->right!=nullptr){
    //                 q.push(node->right);
    //             }
    //         }
    //         level++;
    //     }
    //     return level;
    stack<pair<TreeNode*,int>> stack;
    if(root!=nullptr){
        stack.push({root,1});
    }
    int res=0;
    while(!stack.empty()){
        pair<TreeNode*,int> current=stack.top();
        stack.pop();
        TreeNode* node=current.first;
        int depth=current.second;
        res=max(res,depth);
        if(node->left!=nullptr){
            stack.push({node->left,depth+1});
        }
        if(node->right!=nullptr){
            stack.push({node->right,depth+1});
        }
    }
    return res;
    }
};




